common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/EoClientCertAuthChallengeHandler.java

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.ui.webview.challengehandlers;
+
+import android.annotation.TargetApi;
+import android.app.Activity;
+import android.os.Build;
+import android.security.KeyChain;
+import android.security.KeyChainAliasCallback;
+import android.security.KeyChainException;
+import android.util.Log;
+import android.webkit.ClientCertRequest;
+
+import androidx.annotation.NonNull;
+
+import com.microsoft.identity.common.logging.Logger;
+import com.yubico.yubikit.android.YubiKitManager;
+import com.yubico.yubikit.android.transport.usb.UsbConfiguration;
+import com.yubico.yubikit.android.transport.usb.UsbYubiKeyDevice;
+import com.yubico.yubikit.core.util.Callback;
+
+import java.security.Principal;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+
+public class EoClientCertAuthChallengeHandler implements IChallengeHandler<ClientCertRequest, Void> {
+    private static final String TAG = EoClientCertAuthChallengeHandler.class.getSimpleName();
+    private static final String ACCEPTABLE_ISSUER = "CN=MS-Organization-Access";
+    private Activity mActivity;
+    private final YubiKitManager mYubiKitManager;
+
+    public EoClientCertAuthChallengeHandler(@NonNull final Activity activity) {
+        mActivity = activity;
+        //Create and start YubiKitManager
+        mYubiKitManager = new YubiKitManager(mActivity.getApplicationContext());
+        mYubiKitManager.startUsbDiscovery(new UsbConfiguration(), new Callback<UsbYubiKeyDevice>() {
```

Comment:
Do we need to be discovering USB devices for the entire lifetime or this object or for the entire duration of the WebView being open? Or should we only be discovering when there is actually a client cert request?

Reply:
I did try to just start discovery in the processChallenge method initially, but I noticed that the YubiKey wasn't being detected at all in time to choose a certificate. Allowing discovery for the duration of the WebView being open seems to be more effective at the moment. Soldiers will be plugging in their keys early, most likely at some point when the WebView is showing, and maybe they want a notification telling them that it's been connected as soon as they connect it. But containing the discovery to the challengehandler is something I will still explore as the structure of the EoClientCertAuthChallengeHandler builds.

Reply:
> Maybe calling startDiscovery in the onClientCertRequest method would give enough time for the key to be recognized?
Let's try this out.

> Soldiers will be plugging in their keys early, most likely at some point when the WebView is showing, and maybe they want a notification telling them that it's been connected as soon as they connect it.
If our code sending a System notification? Or is the YubiKey sdk doing it? I didn't realize a notification was being sent.

Reply:
When in discovery mode, the YubiKit SDK shows an alertDialog when a YubiKey is plugged in, asking if it's ok for the host or broker app to access the key. If the user selects ok, then we have access to the key's contents. If they select cancel, the sdk ignores it and does not recognize it as a yubikey anymore.

Reply:
If this is the case then I think the current model of early discovery makes sense.

Reply:
Attempt to start discovery right away in onReceivedClientCertRequest resulted in the same outcome as starting discovery in processChallenge. I think there might be race conditions involved, since I can plug in my yubikey and get the alertDialog after canceling the regular CBA flow.

Reply:
I think the way you have the code currently is good. Now that you explained the notification scenario I think we should keep the code as is. I'll resolve this comment.